src/merkle/merkle_tree.rs

```diff
@@ -19,31 +20,84 @@ impl MerkleTree {
     ///
     /// # Errors
     /// Returns an error if the number of leaves is smaller than two or is not a power of two.
-    pub fn new(leaves: Vec<Word>) -> Result<Self, MerkleError> {
+    pub fn new<T, I, V>(leaves: T) -> Result<Self, MerkleError>
+    where
+        T: IntoIterator<IntoIter = I>,
+        I: ExactSizeIterator + Iterator<Item = V>,
+        V: Borrow<Word>,
+    {
+        let leaves = leaves.into_iter();
         let n = leaves.len();
         if n <= 1 {
             return Err(MerkleError::DepthTooSmall(n as u8));
         } else if !n.is_power_of_two() {
             return Err(MerkleError::NumLeavesNotPowerOfTwo(n));
         }
 
-        // create un-initialized vector to hold all tree nodes
-        let mut nodes = unsafe { uninit_vector(2 * n) };
-        nodes[0] = [Felt::ZERO; 4];
-
-        // copy leaves into the second part of the nodes vector
-        nodes[n..].copy_from_slice(&leaves);
-
-        // re-interpret nodes as an array of two nodes fused together
-        // Safety: `nodes` will never move here as it is not bound to an external lifetime (i.e.
-        // `self`).
-        let ptr = nodes.as_ptr() as *const [RpoDigest; 2];
-        let pairs = unsafe { slice::from_raw_parts(ptr, n) };
+        // The tree contains `n` leaves, which is a power of two. The depth of the tree is `d`
+        // defined as `2^d=n` or `d=lb(n)`. The total number of elements in the tree is
+        // `2**(d+1)-1` or `2n-1` (since `n=2**d`).
+        //
+        // Below we compute the tree size `+1`, the first element is set to `0`, so that the length
+        // of `nodes` will be a power of two.
+        let final_capacity = 2 * n;
+
+        // Allocate data to accomodate all the elements, and get a reference to the underlying
+        // buffer
+        let mut nodes: Vec<Word> = Vec::with_capacity(final_capacity);
+        let buffer = nodes.spare_capacity_mut();
+
+        // The bottom layer goes to the higher indeces (at the end of the vector)
+        let last_layer = &mut buffer[n..];
+        for (pos, el) in leaves.enumerate() {
+            last_layer[pos].write(*el.borrow());
+        }
```

Comment:
Should we expect a performance degradation here as we are moving from `copy_from_slice` (`memcpy`?) to iteration and write per element?

Reply:
I _think_ I found a solution. But it requires a nightly feature [`maybe_uninit_write_slice`](https://doc.rust-lang.org/core/mem/union.MaybeUninit.html#method.write_slice).

The code would then instead be:

```rust
    pub fn new<T>(leaves: T) -> Result<Self, MerkleError>
    where
        T: AsRef<[Word]>,
    {
       ...
       MaybeUninit::write_slice(last_layer, leaves);
```

Reply:
This particular copy doesn't seem to matter. I re-ran the benchmark with the code above and got these results (base is the code in `next`):

```
MerkleTree/construction/65536
                        time:   [317.32 ms 317.77 ms 318.21 ms]
                        change: [+0.6008% +0.8147% +1.0319%] (p = 0.00 < 0.05)
                        Change within noise threshold.
MerkleTree/construction/131072
                        time:   [639.10 ms 639.64 ms 640.17 ms]
                        change: [+0.9486% +1.1657% +1.3779%] (p = 0.00 < 0.05)
                        Change within noise threshold.
MerkleTree/construction/262144
                        time:   [1.2799 s 1.2807 s 1.2815 s]
                        change: [+1.3193% +1.4650% +1.6205%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) low mild
```